### PR TITLE
Remove JSNativeApi.cs

### PR DESCRIPTION
**Pull request by vmoroz**
This PR is a part of the code refactoring that aims to simplify the code and to improve public API.
In this PR we remove the `JSNativeApi.cs` file that mostly had extensions for the `JSValue` struct.
We still have `JSNativeApi` class in the `JSError.cs` file that we can remove in future PR.
The key parts of this PR:
- `JSVativeApi.cs` code is moved inside of the `JSValue` struct.
- Code that access `JSRuntime` functions in `JSValue` is changed to minimize required verifications. New private functions 
  `GetRuntime` and `GetCurrentRuntime` return `napi_env` and `napi_value` in one call.
- Code that was dependent on the `JSNativeApi` is changed to use the new `JSValue` functions.
- `JSMarshaller` changes reflect the function changes from `JSNativeApi` static methods to `JSValue` instance methods.
- Some comment and spelling fixes.

In the next PRs we should improve the `JSValue` API and API for other classes and structs. 
E.g. we can move some methods to JS type specific structs.

### CODE REVIEW

File: src/NodeApi/JSValue.cs [hunk 1]
```diff
@@ -302,27 +282,1121 @@ public static JSValue CreateDataView(int length, JSValue arrayBuffer, int byteOf
...
+            return null;
+        }
+
+        status.ThrowIfFailed();
+        return GCHandle.FromIntPtr(result).Target!;
+    }
+
+    public JSReference CreateReference() => new(this);
+
+    public JSReference CreateWeakReference() => new(this, isWeak: true);
```
Comment (jasongin) [inline src/NodeApi/JSValue.cs:877]: Should these 2 methods be combined into `CreateReference(bool isWeak = false)`?

**Review by jasongin — APPROVED** [review]
Overall this is a great improvement to the public API surface.

I just have some concerns about keeping conversion APIs for `JSBigInt` consistent with other types.

File: src/NodeApi/JSBigInt.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    public static explicit operator JSBigInt?(JSValue value) => value.AsJSBigInt();
+    public static explicit operator JSBigInt(JSValue value)
+        => value.AsJSBigInt() is JSBigInt result
+            ? result
+            : throw new InvalidCastException("JSValue is not BigInt");
+
+    public static implicit operator JSBigInt(BigInteger value) => new(value);
+    public static explicit operator BigInteger(JSBigInt value) => value.ToBigInteger();
+
+    private JSBigInt(JSValue value)
```
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSBigInt.cs:24]: Other types do not have public constructors that receive `JSValue`. They all seemed to be private.
+1 for being consistent.

File: src/NodeApi/JSValue.cs (hunk 1 above)
Comment (vmoroz) [reply src/NodeApi/JSValue.cs:877]: We can add the optional `isWeak` parameter to `CreateReference` but having the separate `CreateWeakReference` is great for the API discoverability.

Comment (jasongin) [reply src/NodeApi/JSValue.cs:877]: The `JSReference` constructor has a `bool isWeak = false` parameter. I'd prefer to be consistent with that, and an additional `CreateWeakReference()` just seems redundant. If you're concerned about discoverability, maybe the parameter should not be optional.

File: src/NodeApi/JSBigInt.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasongin) [reply src/NodeApi/JSBigInt.cs:24]: > Other types do not have public constructors that receive JSValue. They all seemed to be private.

Ahh you're right. However the explicit conversion operators are public, and those are currently unchecked.

File: src/NodeApi/JSValue.cs (hunk 1 above)
Comment (vmoroz) [reply src/NodeApi/JSValue.cs:877]: Should we remove these methods from `JSValue`?

Comment (jasongin) [reply src/NodeApi/JSValue.cs:877]: Probably - they are not any more convenient than the `JSReference` constructor.

File: src/NodeApi/JSBigInt.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+public readonly struct JSBigInt : IEquatable<JSValue>
+{
+    private readonly JSValue _value;
+
+    public static implicit operator JSValue(JSBigInt value) => value.AsJSValue();
+    public static explicit operator JSBigInt?(JSValue value) => value.AsJSBigInt();
+    public static explicit operator JSBigInt(JSValue value)
+        => value.AsJSBigInt() is JSBigInt result
+            ? result
```
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSBigInt.cs]: This code is removed in the latest iteration.

File: src/NodeApi/JSBigInt.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSBigInt.cs:20]: `JSBigInt` is changed to match the API for other types.

File: src/NodeApi/JSValue.cs [hunk 4]
```diff
@@ -302,27 +282,1121 @@ public static JSValue CreateDataView(int length, JSValue arrayBuffer, int byteOf
...
+
+    public bool IsError() => GetRuntime(out napi_env env, out napi_value handle)
+        .IsError(env, handle, out bool result).ThrowIfFailed(result);
+
+    public static bool IsExceptionPending() => GetCurrentRuntime(out napi_env env)
+        .IsExceptionPending(env, out bool result).ThrowIfFailed(result);
+
+    public static JSValue GetAndClearLastException() => GetCurrentRuntime(out napi_env env)
+        .GetAndClearLastException(env, out napi_value result).ThrowIfFailed(result);
```
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSValue.cs]: They are moved to `JSError`.

File: src/NodeApi/JSValue.cs (hunk 1 above)
Comment (vmoroz) [reply src/NodeApi/JSValue.cs]: They are removed.

File: src/NodeApi/JSError.cs [hunk 5]
```diff
@@ -359,7 +381,7 @@ public void Dispose()
     }
 }
 
-public static partial class JSNativeApi
+public static class JSNativeApi
```
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSError.cs]: It is renamed to `NodeApiStatusExtensions` and moved to NodeApiStatusExtensions.cs

File: src/NodeApi/JSBigInt.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            : throw new InvalidCastException("JSValue is not BigInt");
```
[the remark this answers is not in the document]
Comment (vmoroz) [reply src/NodeApi/JSBigInt.cs]: This code is removed to match other types.

**Pull request merged**